include/rmm/mr/device/system_memory_resource.hpp

```diff
@@ -0,0 +1,187 @@
+   * Doing this check can be expensive, so only large buffer above the given threshold will be
+   * checked.
+   *
+   * @param headroom_size Size of the reserved GPU memory as headroom
+   * @param threshold_size Size of the allocation above which to check for headroom
+   */
+  explicit system_memory_resource(std::optional<std::size_t> headroom_size  = std::nullopt,
+                                  std::optional<std::size_t> threshold_size = std::nullopt)
+  {
+    RMM_EXPECTS(rmm::mr::detail::sam::is_supported(),
+                "System memory allocator is not supported with this hardware/software version.");
+    headroom_size_ = headroom_size.value_or(0);
+    if (headroom_size_ == 0) {
+      // If headroom is not specified, always pass through to malloc.
+      threshold_size_ = threshold_size.value_or(std::numeric_limits<std::size_t>::max());
+    } else {
+      // If headroom is specified, by default we check every allocation.
+      threshold_size_ = threshold_size.value_or(0);
+    }
+  }
+
+  ~system_memory_resource() override                    = default;
+  system_memory_resource(system_memory_resource const&) = default;  ///< @default_copy_constructor
+  system_memory_resource(system_memory_resource&&)      = default;  ///< @default_move_constructor
+  system_memory_resource& operator=(system_memory_resource const&) =
+    default;  ///< @default_copy_assignment{system_memory_resource}
+  system_memory_resource& operator=(system_memory_resource&&) =
+    default;  ///< @default_move_assignment{system_memory_resource}
+
+ private:
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * The stream argument is ignored.
+   *
+   * @param bytes The size of the allocation
+   * @param stream This argument is ignored
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, [[maybe_unused]] cuda_stream_view stream) override
+  {
+    auto const aligned{rmm::align_up(bytes, rmm::CUDA_ALLOCATION_ALIGNMENT)};
+    void* ptr{malloc(aligned)};
+
+    if (bytes >= threshold_size_) {
+      auto const free        = rmm::available_device_memory().first;
+      auto const allocatable = std::max(free - headroom_size_, 0UL);
+      auto const gpu_portion = std::min(allocatable, aligned);
+      auto const cpu_portion = aligned - gpu_portion;
+      if (gpu_portion != 0) {
+        RMM_CUDA_TRY(cudaMemAdvise(ptr,
+                                   gpu_portion,
+                                   cudaMemAdviseSetPreferredLocation,
+                                   rmm::get_current_cuda_device().value()));
+      }
+      if (cpu_portion != 0) {
+        RMM_CUDA_TRY(cudaMemAdvise(static_cast<char*>(ptr) + gpu_portion,
+                                   cpu_portion,
+                                   cudaMemAdviseSetPreferredLocation,
+                                   cudaCpuDeviceId));
```

Comment:
*question*: What do we want this interface to look like in a post cuda-12.2 world where `cudaMemAdvice_v2` exists and we can specify numa affinity?

Reply:
Not sure. At least in GH200, there is only one NUMA node on the CPU side, so it doesn't really matter.